### ISSUE TITLE
GenFds should support FDF file sections in any order (Bugzilla Bug 141)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#6656]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 141
Date: 2016-10-07T12:01:00+00:00
From: @mdkinney
To:   @mdkinney
CC:   hot.tian, @lersek, ye.y.wu

Depends on:   #142
Last updated: 2016-10-28T02:32:47+00:00

**Comment by tianocore-issues** [comment]
## Comment 567
Date: 2016-10-07 12:01:00 +0000
From: @mdkinney

- Industry Specification: ---
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @mdkinney

The FDF parse in GenFds current requires sections to be placed in the FDF file in a fixed order.  As the FDF syntax has been extended, this restriction makes it difficult for a platform developer to organize FDF files in a way that makes sense for a specific platform.  

The current fixed order is: Fd, Fv, Fmp, Capsule, Vtf, Rule, OptionRom

The EDK II FDF Specification does currently require a specific order of sections.

Another issue with the current implementation is that sections that do not appear in the fixes order are silently ignored and may generate build breaks with no information on why the build broke.  Even worse, the build may complete with no errors, but does not produce content associated with the sections that were skipped.

This issue was discovered during evaluation of FMP based system firmware updates when attempting to interleave [FmpPayload] and [Capsule] sections for firmware update and recovery on Galileo platforms. 

The proposed fix is to relax the FDF parser in GenFds to allow sections to appear in any order and collect the information from all valid section not matter what order they appear in the FDF file.

**Comment by tianocore-issues** [comment]
## Comment 570
Date: 2016-10-07 12:05:27 +0000
From: @mdkinney

EDK II FDF Specification update is at:

https://bugzilla.tianocore.org/show_bug.cgi?id=142

**Comment by tianocore-issues** [comment]
## Comment 571
Date: 2016-10-07 12:21:53 +0000
From: @lersek

See also the report / discussion in the following edk2-devel thread:

[edk2] Question on OvmfPkgX64.fdf
https://www.mail-archive.com/edk2-devel@lists.01.org/msg02183.html

**Comment by tianocore-issues** [comment]
## Comment 588
Date: 2016-10-12 21:07:28 +0000
From: Chris &lt;<ye.y.wu>&gt;

assign to Mike

**Comment by tianocore-issues** [comment]
## Comment 789
Date: 2016-10-28 02:32:47 +0000
From: @mdkinney

https://github.com/tianocore/edk2/commit/dd170333f6444a4256e75356a8f0758a40bfb77d
